# Proyecto Final

## Integrantes

- Sergio Orellana - 221122
- Andy Fuentes - 22944
- Christian Echeverria -
- Jorge Lopez -
- Davis Roldan -


## Análisis exploratorio

En este notebook realizaremos:

1. **Evaluación de la calidad de datos**: Revisión de dimensiones, tipos de datos, valores nulos y duplicados de cada CSV.
2. **Resúmenes estadísticos**: Análisis descriptivo (medias, desviaciones, percentiles, etc.) de las variables numéricas.
3. **Visualización de patrones clave**: Gráficos para identificar distribuciones y relaciones relevantes.
4. **Análisis de correlación de características**: Exploración de relaciones entre variables numéricas en datasets individuales y combinados.

Asegúrate de tener instaladas las librerías necesarias: `pandas`, `numpy`, `matplotlib` y `seaborn`.


### Importar librerías


In [9]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import io

### Carga de los datasets


In [7]:
df_customers = pd.read_csv("./Data/olist_customers_dataset.csv")
df_geolocation = pd.read_csv("./Data/olist_geolocation_dataset.csv")
df_order_items = pd.read_csv("./Data/olist_order_items_dataset.csv")
df_order_payments = pd.read_csv("./Data/olist_order_payments_dataset.csv")
df_order_reviews = pd.read_csv("./Data/olist_order_reviews_dataset.csv")
df_orders = pd.read_csv("./Data/olist_orders_dataset.csv")
df_products = pd.read_csv("./Data/olist_products_dataset.csv")
df_sellers = pd.read_csv("./Data/olist_sellers_dataset.csv")
df_prod_cat_trans = pd.read_csv("./Data/product_category_name_translation.csv")

### Evaluación de la calidad de datos

Revisamos las siguientes cuestiones:

- Dimensiones (número de filas y columnas)
- Información general y tipos de datos
- Valores nulos por columna
- Registros duplicados

A continuación, se realiza este chequeo para cada dataset.


In [12]:
datasets = {
    "olist_customers_dataset": df_customers,
    "olist_geolocation_dataset": df_geolocation,
    "olist_order_items_dataset": df_order_items,
    "olist_order_payments_dataset": df_order_payments,
    "olist_order_reviews_dataset": df_order_reviews,
    "olist_orders_dataset": df_orders,
    "olist_products_dataset": df_products,
    "olist_sellers_dataset": df_sellers,
    "product_category_name_translation": df_prod_cat_trans,
}

for name, df in datasets.items():
    print(f"### {name}")
    print("Dimensiones:", df.shape)
    print("\nInformación:")
    print(df.info())
    print("\nValores nulos por columna:")
    print(df.isnull().sum())
    print("Registros duplicados:", df.duplicated().sum())
    print("\n" + "=" * 70 + "\n")

### olist_customers_dataset
Dimensiones: (99441, 5)

Información:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None

Valores nulos por columna:
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
Registros duplicados: 0


### olist_geolocation_dataset
Dimensiones: (1000163, 5)

Información:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 colu

1. **olist_customers_dataset**

- Dimensiones y Estructura:
  Contiene 99,441 registros y 5 columnas.

- Contenido y Calidad:
  Todas las filas están completas (sin valores nulos) para las columnas de identificación, código postal, ciudad y estado.

- Observaciones:
  El dataset es muy consistente, ya que no presenta duplicados ni datos faltantes. Esto indica que la información de los clientes (incluyendo el identificador único y la ubicación) es confiable y se puede utilizar directamente para realizar análisis demográficos o para vincular con otros datasets (por ejemplo, al hacer joins con pedidos).

2. **olist_geolocation_dataset**

- Dimensiones y Estructura:
  Contiene 1,000,163 registros repartidos en 5 columnas (prefijo de código postal, latitud, longitud, ciudad y estado).

- Contenido y Calidad:
  No presenta valores nulos; sin embargo, se detectaron 261,831 registros duplicados.

- Observaciones:
  La alta cantidad de duplicados sugiere que existen múltiples entradas para algunos códigos postales. Esto puede ser deliberado (por ejemplo, diversas coordenadas o ubicaciones que corresponden al mismo código postal) o podría requerir una depuración, especialmente si el objetivo es generar una lista única de ubicaciones o realizar análisis geoespaciales precisos.

3. **olist_order_items_dataset**

- Dimensiones y Estructura:
  Consta de 112,650 registros y 7 columnas, que incluyen identificadores de órdenes y productos, detalles del vendedor, fecha límite de envío, precio y valor del frete.

- Contenido y Calidad:
  No hay datos nulos ni duplicados en este conjunto; los campos numéricos (como precio y valor del frete) tienen el tipo de dato adecuado (float64).

- Observaciones:
  Este dataset está limpio y preparado para realizar análisis de costos, distribución de precios y frete, y detectar posibles outliers en las transacciones.

4. **olist_order_payments_dataset**

- Dimensiones y Estructura:
  Consta de 103,886 registros y 5 columnas, incluyendo el identificador de la orden, secuencia de pago, tipo de pago, número de cuotas y valor de pago.

- Contenido y Calidad:
  No se detectan valores nulos ni duplicados.

- Observaciones:
  La calidad de estos datos es alta y permite realizar análisis sobre métodos de pago, distribución de cuotas y correlación entre forma de pago y valor de transacción. Es importante notar la diferencia en el número de registros respecto a otros datasets, lo que puede indicar que algunas órdenes pueden tener más de un registro en el contexto de pagos.

5. **olist_order_reviews_dataset**

- Dimensiones y Estructura:
  Incluye 99,224 registros y 7 columnas relacionadas con evaluaciones de las órdenes.

- Contenido y Calidad:

Los campos críticos como review_id, order_id, review_score, review_creation_date y review_answer_timestamp están completos.

Las columnas de comentarios (review_comment_title y review_comment_message) tienen un alto porcentaje de valores nulos (por ejemplo, solo alrededor de 11,568 registros tienen título y 40,977 contienen mensaje), lo que puede reflejar que no todos los clientes dejan retroalimentación textual.

- Observaciones:
  Aunque la puntuación numérica (score) es íntegra, la información cualitativa (comentarios) es esporádica, lo que sugiere que para análisis de sentimiento o de texto se trabajará únicamente con la información disponible o se aplicarán técnicas para tratar la ausencia de estos datos.

6. **olist_orders_dataset**

- Dimensiones y Estructura:
  Contiene 99,441 registros y 8 columnas relacionadas con el ciclo completo de una orden (compra, aprobación y fechas de entrega).

- Contenido y Calidad:

La mayoría de las columnas están completas; sin embargo, existen algunos valores nulos en columnas de fechas:

order_approved_at: 160 nulos.

order_delivered_carrier_date: 1,783 nulos.

order_delivered_customer_date: 2,965 nulos.

Esto es común en escenarios donde algunas órdenes pueden no haberse completado o existir retrasos en la actualización.

- Observaciones:
  El dataset es esencial para analizar el ciclo de vida de los pedidos, y los campos con valores nulos se deben revisar para entender qué significan en el contexto del negocio (por ejemplo, pedidos cancelados o retrasados). Las fechas se deben convertir a formato datetime para facilitar el análisis temporal.

7. **olist_products_dataset**

- Dimensiones y Estructura:
  Consta de 32,951 registros y 9 columnas, incluyendo identificadores de producto, detalles físicos y de la descripción.

- Contenido y Calidad:

La mayoría de las columnas tienen datos casi completos, salvo en columnas relacionadas con detalles descriptivos (como product_category_name, product_name_lenght, product_description_lenght y product_photos_qty), donde se observan aproximadamente 610 registros nulos (alrededor del 1.8-2% de los registros).

Los atributos físicos (peso, dimensiones) tienen muy pocos valores nulos, lo que sugiere que la mayoría de los productos cuentan con información detallada de sus características.

- Observaciones:
  Este dataset es crucial para clasificar productos y relacionar características físicas con el tipo de producto. La poca cantidad de datos faltantes en campos clave permite un análisis robusto y confiable.

8. **olist_sellers_dataset**

- Dimensiones y Estructura:
  Contiene 3,095 registros y 4 columnas que incluyen el identificador del vendedor, el código postal, la ciudad y el estado.

- Contenido y Calidad:
  No se reportan valores nulos ni duplicados.

- Observaciones:
  La información de los vendedores es completa y puede utilizarse para evaluar el impacto de la localización o la reputación de un vendedor en la logística y tiempos de entrega. Es ideal para vincular con otros datos (por ejemplo, dentro del análisis de pedidos o envío).

9. **product_category_name_translation**

- Dimensiones y Estructura:
  Este dataset es relativamente pequeño con 71 registros y 2 columnas (nombre de la categoría en portugués y su traducción al inglés).

- Contenido y Calidad:
  Todas las entradas están completas sin valores nulos ni duplicados.

- Observaciones:
  Este archivo es fundamental para la interpretación de categorías de producto en contextos internacionales o para reportes en inglés, permitiendo una estandarización de la nomenclatura en análisis posteriores.


### Resúmenes Estadísticos

En esta sección haremos resúmenes estadísticos para cada dataset utilizando el método `.describe()`.

Se muestran:

- **Resumen numérico:** Estadísticas básicas (conteo, media, desviación estándar, mínimo, percentiles y máximo) para columnas numéricas.
- **Resumen categórico:** Estadísticas descriptivas (conteo, número de valores únicos, valor más frecuente y su frecuencia) para columnas de tipo objeto.

Esto nos ayudará a identificar la dispersión de las variables, detectar posibles outliers y entender la distribución de las categorías.


In [13]:
def show_summary_stats(df, dataset_name):
    print(f"### {dataset_name}")

    # Resumen numérico: para variables numéricas
    print("\n#### Resumen numérico:")
    print(df.describe())

    # Resumen categórico: para variables de tipo object (cadenas de texto)
    print("\n#### Resumen categórico:")
    print(df.describe(include=["object"]))

    print("\n" + "=" * 70 + "\n")


# Diccionario de datasets (asegúrate de haber cargado previamente los CSV)
datasets = {
    "olist_customers_dataset": df_customers,
    "olist_geolocation_dataset": df_geolocation,
    "olist_order_items_dataset": df_order_items,
    "olist_order_payments_dataset": df_order_payments,
    "olist_order_reviews_dataset": df_order_reviews,
    "olist_orders_dataset": df_orders,
    "olist_products_dataset": df_products,
    "olist_sellers_dataset": df_sellers,
    "product_category_name_translation": df_prod_cat_trans,
}

# Iteramos sobre cada dataset y mostramos los resúmenes
for name, df in datasets.items():
    show_summary_stats(df, name)

### olist_customers_dataset

#### Resumen numérico:
       customer_zip_code_prefix
count              99441.000000
mean               35137.474583
std                29797.938996
min                 1003.000000
25%                11347.000000
50%                24416.000000
75%                58900.000000
max                99990.000000

#### Resumen categórico:
                             customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_city customer_state  
count          99441          99441  
unique          4119             27  
top        sao paulo             SP  
freq           15540          41746  


### olist_geolocation_dataset

#### Resumen numéric